File: libs/python/deploy/shadow_verify.py
"""On-device verification (Python) of the sanitized default SHADOW name.

Run as a Greengrass component with `-c SHADOW` (no explicit name): the SHADOW config
manager defaults the shadow name to the component name and sanitizes it
(``com.ggcommons.PyShadowVerify`` -> ``com_ggcommons_PyShadowVerify``), then loads
config from that named shadow via ShadowManager IPC. ``get_config_source()`` reports
the resolved shadow name, and the loaded marker values (set in the cloud shadow under
the sanitized name) prove the default->sanitize->GetThingShadow path runs end-to-end.
"""
import json
import sys

from ggcommons import GGCommonsBuilder

COMPONENT = "com.ggcommons.PyShadowVerify"
RESULT = "/tmp/python_shadow_verify_result.json"


def main():
    out = {"lang": "python"}
    try:
        gg = (
            GGCommonsBuilder.create(COMPONENT)
            .with_args(sys.argv[1:])
            .receive_own_messages(False)
            .build()
        )
        cm = gg.get_config_manager()
        out["connected"] = True
        out["config_source"] = cm.get_config_source()  # includes the resolved shadow name
        out["config_loaded"] = {
            "publish_interval": (cm.get_global_config() or {}).get("publish_interval"),
            "site": (cm.get_full_config().get("tags") or {}).get("site"),
            "thing": cm.get_thing_name(),
        }
        with open(RESULT, "w") as f:
            json.dump(out, f)
        gg.shutdown()
    except Exception as e:  # noqa: BLE001 - harness records any failure
        out["connected"] = False
        out["error"] = str(e)
        with open(RESULT, "w") as f:
            json.dump(out, f)
        sys.exit(1)


if __name__ == "__main__":
    main()

File: libs/java/deploy/shadow-verify/src/main/java/com/mbreissi/edgecommons/shadowverify/ShadowVerify.java
package com.mbreissi.edgecommons.shadowverify;

import com.mbreissi.edgecommons.EdgeCommons;
import com.mbreissi.edgecommons.EdgeCommonsBuilder;
import com.mbreissi.edgecommons.config.ConfigManager;
import com.google.gson.JsonObject;

import java.nio.file.Files;
import java.nio.file.Paths;

/**
 * On-device verification (Java) of the sanitized default SHADOW name.
 *
 * Run as a Greengrass component with {@code -c SHADOW} (no explicit name): the SHADOW
 * config provider defaults the shadow name to the component name and sanitizes it
 * ({@code com.mbreissi.edgecommons.JavaShadowVerify} -> {@code com_mbreissi_edgecommons_JavaShadowVerify}),
 * then loads config from that named shadow via ShadowManager IPC. Loading the marker
 * values (set in the cloud shadow under the sanitized name) proves the
 * default->sanitize->GetThingShadow path runs end-to-end.
 */
public final class ShadowVerify {
    private static final String COMPONENT = "com.mbreissi.edgecommons.JavaShadowVerify";
    private static final String RESULT = "/tmp/java_shadow_verify_result.json";

    public static void main(String[] args) {
        try {
            EdgeCommons gg = EdgeCommonsBuilder.create(COMPONENT).withArgs(args).build();
            ConfigManager cm = gg.getConfigManager();

            JsonObject global = cm.getGlobalConfig();
            String publishInterval = (global != null && global.has("publish_interval"))
                    ? global.get("publish_interval").getAsString() : "null";
            String site = (cm.getTagConfig() != null) ? cm.getTagConfig().getKeyValue("site") : null;

            String json = String.format(
                    "{\"lang\":\"java\",\"connected\":true,\"config_loaded\":"
                            + "{\"publish_interval\":%s,\"site\":\"%s\",\"thing\":\"%s\"}}",
                    publishInterval, site, cm.getThingName());
            Files.write(Paths.get(RESULT), json.getBytes());

            Thread.sleep(20000); // stay RUNNING briefly, then exit
            System.exit(0);
        } catch (Exception e) {
            try {
                Files.write(Paths.get(RESULT),
                        ("{\"lang\":\"java\",\"connected\":false,\"error\":\""
                                + e.toString().replace("\"", "'") + "\"}").getBytes());
            } catch (Exception ignored) {
                // best effort
            }
            System.exit(1);
        }
    }
}
